### ISSUE TITLE
Interpolation loop

### ISSUE THREAD

**Issue by Peter9192** [WAM2layers/WAM2layers#20]
Half of the time in the preprocessing is spent on interpolation of the pressure level data to a denser grid. The current implementation is suboptimal, since python loops are slow:

https://github.com/WAM2layers/WAM2layers/blob/c97a8281a3db3dfaa0873dca3bc9e49496d62e4b/Fluxes_and_States_mat.py#L241-L261

The challenge is that the interpolation doesn't go from one fixed grid to another fixed grid, but that the source grid (pressure field) is variable in space and time. 

Some possibilities for optimization:

- Use a library that is able to map from a variable to a fixed pressure grid. 
  - Can [xarray's advanced indexing](https://xarray.pydata.org/en/stable/user-guide/interpolation.html#advanced-interpolation) be used for this?
  - GeoCAT (successor of NCL) seems to have "interp_hybrid_to_pressure" and "interp_sigma_to_pressure" [here](https://github.com/NCAR/geocat-comp/blob/993bcbf9e642ed9e948e1c0f6087d3b882a515f2/src/geocat/comp/interpolation.py#L51)
  - metpy? https://unidata.github.io/MetPy/latest/examples/sigma_to_pressure_interpolation.html
- Use a custom interpolation routine. We (@Imme1992 and I) once worked on an interpolation function that only requires a single pass over the data, but there might still be some issues with it. See https://stackoverflow.com/a/49736163/6012085 and https://github.com/Peter9192/Python/blob/master/Interpolation_update.ipynb
- Implement the nested loops in a low-level language and make that available as a Python function (use cython).
- Use an external tool/package/language (possibly call from within python?). 
  - It seems [NCL has some functions](https://www.ncl.ucar.edu/Applications/vert_interp.shtml). Note that NCL will be replaced by GeoCAT, see above
  - Perhaps CDO or nco has something 
  - Regridding from sigma/hybrid levels to pressure levels is a common enough operation that there should be some tools around.

**Comment by chrisweij** [comment]
Hi Peter, 

CDO only has the option to interpolate from model to pressure levels (function ml2pl), which I have used and which works pretty well but is probably the opposite what you want. 

NCO has the ncremap program (http://nco.sourceforge.net/nco.html#Options-specific-to-ncremap), which has the --vrt_fl option which might be a good option, but I have not used that before. 

The MetPy packages has also some (vertical) interpolation functions, but I do not think they have specific functions for interpolating to model levels. 

On another note, would it not be possible to directly interpolate it from pressure levels to the 2 layers used by WAM2layers, instead of interpolate it to model levels as an intermediate step? (Given that you take care of that data on pressure levels should be set to nan if the pressure level is below the surface pressure).

**Comment by Peter9192** [comment]
Thanks for the suggestions! I don't think we're interpolating to model levels, just to a denser grid so as to have a better representation of the vertical structure before splitting it in 2 layers. I think @Imme1992 can probably explain better why this is preferred over directly interpolating to the 2 layers. But if possible, I agree it would be easier to skip it.

I also realized that the original data, in this case, are probably already (some) model levels, not pressure levels. Sorry for the confusion about that.

**Comment by Imme1992** [comment]
Yes indeed, I checked the code again, the data is not interpolated to model levels, but to a more dense grid of pressure levels. To take into account features as a low level jet etc, we used a spline interpolation to go from 5 pressure levels to 40 pressure levels, which provided more realistic vertical profiles of u*q and v*q compared to a linear interpolation (I compared with profiles of u*q and v*q from model levels).
See the figure below with a vertical profile of u*q and v*q for many model levels, 5 pressure levels, 5 pressure levels linearly interpolated to 40 pressure levels, and 5 pressure levels spline interpolated to 40 pressure levels. This is figure 2.2 of my PhD thesis.

![image](https://user-images.githubusercontent.com/18259095/159674322-9069c486-1a03-411e-b30b-b6a43e543345.png)

However you could wonder if this extra (and computationally time consuming step) is worth it (we could test that), and if we should not directly aggregate to two layers. So what is the difference if we aggregate directly to two layers, the problem is then that you sometimes only have one value for the bottom layer (only surface pressure for example), as the next layer is 850 hPa, which can also relate to the top layer (have to check what is the average boundary pressure level between the two layers).

In the original code, loading in the data at model levels, the conversion to pressure levels is also directly made. We could consider applying an algorithm as suggested to make that faster

**Comment by Peter9192** [comment]
Nice figures! It would be nice if we can move the interpolation code to a separate file/function, so it can easily be exchanged with other options.

**Comment by Peter9192** [comment]
Perhaps this package might also be useful: https://xgcm.readthedocs.io/en/latest/transform.html

**Comment by Imme1992** [comment]
This is currently not necessary because we have an alternative implementation for dealing with (ERA5) pressure levels. If we get another case where we have fewer pressure levels, we might need to re-introduce this functionality.

Alternative approach could be to solve this with numba.
